tests/Database/DatabaseQueryBuilderTest.php

```diff
@@ -1090,6 +1090,34 @@ public function testUpdateMethodRespectsRaw()
         $this->assertEquals(1, $result);
     }
 
+    public function testUpdateOrInsertMethod()
+    {
+        $builder = m::mock('Illuminate\Database\Query\Builder[where,first,insert]', [
+            m::mock('Illuminate\Database\ConnectionInterface'),
+            new Illuminate\Database\Query\Grammars\Grammar,
+            m::mock('Illuminate\Database\Query\Processors\Processor'),
+        ]);
+
+        $builder->shouldReceive('where')->once()->with(['email' => 'foo'])->andReturn(m::self());
+        $builder->shouldReceive('first')->once()->andReturn(null);
+        $builder->shouldReceive('insert')->once()->with(['email' => 'foo', 'name' => 'bar'])->andReturn(true);
+
+        $this->assertEquals(true, $builder->updateOrInsert(['email' => 'foo'], ['name' => 'bar']));
+
+        $builder = m::mock('Illuminate\Database\Query\Builder[where,first,update]', [
+            m::mock('Illuminate\Database\ConnectionInterface'),
+            new Illuminate\Database\Query\Grammars\Grammar,
+            m::mock('Illuminate\Database\Query\Processors\Processor'),
+        ]);
+
+        $builder->shouldReceive('where')->twice()->with(['email' => 'foo'])->andReturn(m::self());
+        $builder->shouldReceive('first')->once()->andReturn(m\Matcher\Any::class);
+        $builder->shouldReceive('take')->andReturnSelf();
+        $builder->shouldReceive('update')->once()->with(['name' => 'bar'])->andReturn(1);
+
+        $this->assertEquals(true, $builder->updateOrInsert(['email' => 'foo'], ['name' => 'bar']));
+    }
+
```

Comment:
The test is concerned about checking if the update or insert methods are called based on the existence of a row, not sure if there's a better way to make such a test.

I'm still learning much looking at the framework tests, tried to find tests for `Eloquent/Builder::updateOrCreate()` but looks like this method is not covered and the closest alternative was `testFindOrFailMethodThrowsModelNotFoundException()` which mocks `first()` and `where()`.

If you can give me a hint about how to better test such method I'd be grateful.